### DISCUSSION TITLE
How can I get mae and rmse of validation and training

### DISCUSSION THREAD

**Discussion by turbosonics** [ACEsuit/mace#379] [Q&A] [answered]
Hello, 

I hope to check mae and rmse of E, F, and stress (and maybe virial too) from validation set and training set. But I don't know how to set such output option. Following is the MACE input setting I'm using for my tests;
```
mace_run_train \
    --name="Test" \
    --seed=123 \
    --device=cuda \
    --default_dtype="float32" \
    --error_table="PerAtomMAE" \
    --model="MACE" \
    --r_max=6.0 \
    --hidden_irreps='128x0e + 128x1o' \
    --num_channels=32 \
    --max_L=2 \
    --compute_stress=True \
    --compute_forces=True \
    --train_file="${GEOMETRY_DIR}/${GEOMETRY_TRAIN_NAME}" \
    --valid_fraction=0.3 \
    --test_file="${GEOMETRY_DIR}/${GEOMETRY_TEST_NAME}" \
    --E0s="average" \
    --energy_key="energy" \
    --forces_key="forces" \
    --stress_key="stress" \
    --loss="weighted" \
    --forces_weight=1.0 \
    --energy_weight=1.0 \
    --stress_weight=1.0 \
    --config_type_weights='{"Default":1.0}' \
    --optimizer="adam" \
    --batch_size=5 \
    --valid_batch_size=5 \
    --lr=0.005 \
    --amsgrad \
    --scheduler="ReduceLROnPlateau" \
    --lr_factor=0.5 \
    --scheduler_patience=50 \
    --lr_scheduler_gamma=0.99 \
    --ema \
    --ema_decay=0.99 \
    --max_num_epochs=500 \
    --patience=50 \
    --restart_latest \
    --save_cpu \
    --clip_grad=100.0 \
```

And this is what I'm getting from result: 
```
...
{"loss": 0.0002929910842794925, "time": 0.2666587829589844, "mode": "opt", "epoch": 498}
{"loss": 0.002309008482221874, "mae_e": 0.15782718360424042, "mae_e_per_atom": 0.0007970419828779995, "rmse_e": 0.22078144550323486, "rmse_e_per_atom": 0.0011314523871988058, "q95_e": 0.47121582031249964, "mae_f": 0.024527879431843758, "rel_mae_f": 3.2211644996095816, "rmse_f": 0.04782024770975113, "rel_rmse_f": 3.510639968794681, "q95_f": 0.09724280387163162, "mae_stress": 0.0073347738943994045, "rmse_stress": 0.01363541092723608, "rmse_stress_per_atom": 7.09116502548568e-05, "q95_stress": 0.029036216065287588, "mae_virials": 62.1697883605957, "rmse_virials": 129.260009765625, "rmse_virials_per_atom": 0.6706804037094116, "q95_virials": 393.78586730957005, "time": 113.37074255943298, "mode": "eval", "epoch": 498}
{"loss": 0.00036834419006481767, "time": 0.27051448822021484, "mode": "opt", "epoch": 499}
{"loss": 0.00502742687240243, "time": 0.27277469635009766, "mode": "opt", "epoch": 499}
...
```

Is this something related to --error_table option? Or --loss option? 

Thanks

**Comment by pobo95** [comment, marked as answer]
`--error_table=PerAtomRMSEstressvirials ` this option maybe be help you

**Reply by turbosonics** [reply]
Thanks, but I don't see any separate error data from validation set and training set from result file or log file... 

In my test training cases, I'm training with different amount of VASP AIMD data to see if data overfits or underfits with given number of AIMD images, for several different validation set ratio (I usually begins from 20:80 and change according to test results). 

But I couldn't find separate mae and rsme values from training set and validation set. For example, in the output file, 

```
...
{"loss": 0.004128155764192343, "time": 0.26374125480651855, "mode": "opt", "epoch": 2}
{"loss": 0.01074489706061837, "mae_e": 1.0948643684387207, "mae_e_per_atom": 0.005093080922961235, "rmse_e": 1.6352043151855469, "rmse_e_per_atom": 0.007114987820386887, "q95_e": 4.299835205078125, "mae_f": 0.05476377531886101, "rel_mae_f": 7.191943739444998, "rmse_f": 0.10303162038326263, "rel_rmse_f": 7.563886468396652, "q95_f": 0.1979506030678749, "mae_stress": 0.007318621501326561, "rmse_stress": 0.013619007542729378, "rmse_stress_per_atom": 7.068830745993182e-05, "q95_stress": 0.02739041484892368, "mae_virials": 60.9088134765625, "rmse_virials": 130.17013549804688, "rmse_virials_per_atom": 0.6770652532577515, "q95_virials": 392.84836883544864, "time": 114.35933327674866, "mode": "eval", "epoch": 2}
{"loss": 0.03137025237083435, "time": 0.2750709056854248, "mode": "opt", "epoch": 3}
{"loss": 0.0024351708125323057, "time": 0.2458961009979248, "mode": "opt", "epoch": 3}
...
```

I assume these mae and rmse values are validation set ones, right? But where can I find the same rmse and mae values of training set, or how can I print out training set rmse and mae values, so that I can compare curves to see if training goes overfitting or underfitting...

**Reply by pobo95** [reply]
When you check the log file, MACE seperated validation set from training set automatily (maybe 20% of training set). 

When the training is finished (reach to max_ephoch), it shows RMSE or MAE of test set.

But during the training, it shows RMSE/MAE of validation sets.

If you want to see training sets' RMSE you should change code little (MACE/tools/train.py and MACE/modules/loss.py .etc..)

**Reply by turbosonics** [reply]
Please correct me if I'm wrong, but from what I know about graph neural network training in general, we need to compare errors (RMSE and MAE) between validation set and training set to make a decision on overfitting and underfitting. Right? 

I think I'm doing something wrong, but I still don't get it. Why users need to change the code to check RMSE and MAE of loss functions for training set and validation set to judge overfitting and underfitting?... Wouldn't it be better for MACE if the code displays them from the beginning from log and output file?  

Could you let me know how did you decide overfitting or underfitting of given training data set from AIMDs, without modifying codes, from MACE training?

**Reply by ilyes319** [reply]
You can look at the training error in the results folder, it saves a training log for each batch. 

In "modern neural networks" the loss value will go down close to zero for most dataset. The best to monitor overfitting is to check if the validation loss increases.

**Reply by pobo95** [reply]
@turbosonics 
Actually you don't need to check training set's RMSE/MAE.

If overfitting occurs, it cannot be distinguished only by the RMSE/MAE of the training set.
This is because the loss decreases regardless of overfitting. 

However, overfitting can be confirmed with the value of validation as @ilyes319 said.

In case of underfitting
RMSE/MAE of the training set will increase and the value of validation set will also increase.


Therefore, you only need to check the RMSE/MAE of the validation set.
If it increases, something is going wrong.

**Reply by turbosonics** [reply]
@ilyes319 @pobo95 
Thanks for your suggestions. 

I don't have huge amount of GNN model training experience. But from other GNN model training project for different applications (it was not about GNNFF MD), I experienced some cases where loss has declined close to zero and keep nice steady state, yet the model is declared as overfitting because validation set RMSE (and MAE) of a single loss function was clearly higher than training set RMSE. But all other curves were fine. Back then, we concluded the training data had a lot of noise which caused such behavior, so we used several different measures to bypass... 

For case of GNNFF MD framework, I still think such scenarios might be possible (but I might be wrong). We could think of an example scenario of fine-tuning (or from the scratch) training where AIMD training data contains a lot of various configurations such as surface reconstruction, interface behavior, NPT ensembles of high temperature, structural deformation of various crystal and/or non-crystals... 

But if MACE can digest those, then I will be happy! Also, if there is any training strategy/approach that helps to train such AIMD data using MACE, I really hope to learn more about this!  

Thanks! 

PS) Oh, and please forgive my ignorance, but could anyone explain me what is "test set" of MACE means? Is this means the "validation set" of typical other GNN model training? I understand that way but please correct me if I was wrong.

**Comment by gabor1** [comment]
I would certainly hope that MACE can "digest" any reasonable configuration.

**Comment by gabor1** [comment]
Test set is an entirely independent set that you can specify and RMSEs get printed for it, but it is not used in any way in training. Validation set is used to stop the training.

**Comment by nihil39** [comment]
Hi, 

So the error for validation set is in the Log folder while the error for training and test set (if specified) is in the Results folder, right?

**Reply by ilyes319** [reply]
The validation and test errors are in the log file and the train errors and additional metrics on the validation are in the results folder. I recommend looking mostly at the log file.

**Reply by nihil39** [reply]
Thanks, so if I evaluate the model at every epoch I will get the error for the validation set at every epoch in the Result folder, right?
Also in the results folder we have the error for every batch of the train set, right?

An example of what I see in my results file below.

Thanks. 

 ```
{"loss": 0.04125154951473909, "time": 0.7385518550872803, "mode": "opt", "epoch": 0},
                                                                                                     
   {"loss": 0.005330307687914215, "mae_e": 21.192024870650133, "mae_e_per_atom": 0.05651206632173369, "rmse_e": 27.601802432631178, "rmse_e_per_atom": 0.07360480648701648, "q95_e": 51.90141      890064102, "mae_f": 0.16915851764028444, "rel_mae_f": 29.732191547981383, "rmse_f": 0.21882735053267677, "rel_rmse_f": 28.90630629140727, "q95_f": 0.4350285646792365, "mae_stress": 0.064      80060754203766, "rmse_stress": 0.11177974582671243, "rmse_stress_per_atom": 0.0002980793222045665, "q95_stress": 0.2100111132018144, "mae_virials": 252.9917387438205, "rmse_virials": 432      .76730130005774, "rmse_virials_per_atom": 1.1540461368001542, "q95_virials": 775.3135284800139, "time": 2.4479868412017822, "mode": "eval", "epoch": 0},  
                                
  {"loss": 0.043309247581271894, "time": 0.7393889427185059, "mode": "opt", "epoch": 1},   
```

The first line is the last batch of the training set for epoch 0 while the second line should be the error on the validation set, even on last epochs I have similar errors. So, if I understand correctly, this model is predicting better for validation set than for test set?
